File: code_samples/r/functional_closures.r
# Functional: Closures
# AI/ML Training Sample

Closures <- setRefClass(
    "Closures",
    fields = list(data = "character"),
    methods = list(
        initialize = function() {
            data <<- ""
        },
        process = function(input) {
            data <<- input
        },
        getData = function() {
            return(data)
        },
        validate = function() {
            return(nchar(data) > 0)
        }
    )
)

# Example usage
instance <- Closures$new()
instance$process("example")
cat("Data:", instance$getData(), "\n")
cat("Valid:", instance$validate(), "\n")

File: code_samples/ruby/functional_closures.rb
# Functional: Closures
# AI/ML Training Sample

class Closures
  attr_accessor :data
  
  def initialize
    @data = ""
  end
  
  def process(input)
    @data = input
  end
  
  def validate
    !@data.empty?
  end
  
  def to_s
    "Data: #{@data}"
  end
end

# Example usage
instance = Closures.new
instance.process("example")
puts instance
puts "Valid: #{instance.validate}"
